PsfLauncher/PsfPowershellScriptRunner.h

```diff
@@ -285,4 +293,34 @@ class PsfPowershellScriptRunner
 
         return true;
     }
+
+    void MakeAttributeList(LPPROC_THREAD_ATTRIBUTE_LIST& attributeList)
+    {
+        SIZE_T AttributeListSize{};
+
+        InitializeProcThreadAttributeList(nullptr, 1, 0, &AttributeListSize);
+        attributeList = (LPPROC_THREAD_ATTRIBUTE_LIST)HeapAlloc(

```

Comment:
>attributeList = (LPPROC_THREAD_ATTRIBUTE_LIST)HeapAlloc( [](start = 8, length = 56)

Where is the corresponding delete/free call (DeleteProcThreadAttributeList)?

Can we use an RAII wrapper for this type?

e.g.
    using unique_proc_thread_attribute_list = wil::unique_any<LPPROC_THREAD_ATTRIBUTE_LIST, decltype(&DeleteProcThreadAttributeList), DeleteProcThreadAttributeList>;

Reply:
There is none.  

I've never used the wil::unique_any before.  Where would be the best place to put it?

I'm guessing the destructor of the ScriptInformation struct.  That way the attribute list memory will be freed when the structs go out of scope.

Another idea would be either
1. After the CreateProcess call in StartProcessHelper
2. After the StartProcess calls in PowershellScriptHelper.

Unless I'm wrong, I believe it needs to be called only once because the assignments only copy the pointer, and not the data the pointer points to.

Reply:
I think you should design a little struct (e.g. MyProcThreadAttributeList) at the top of this file that owns the memory for the PROC_THREAD_ATTRIBUTE_LIST, owns the memory for the DWORD (because this memory must outlive the PROC_THREAD_ATTRIBUTE_LIST memory) that specifies creation in the container, and is responsible for calling DeleteProcThreadAttributeList.

I think we should avoid using HeapAlloc() - just use new and have a standard c++ unique_ptr manage that memory (but the smart ptr should be owned by the MyProcThreadAttributeList struct you made above).

Does that all work?

---
In reply to: [394060771](https://github.com/microsoft/MSIX-PackageSupportFramework/pull/136#discussion_r394060771) [](ancestors = 394060771)

Reply:
I think this should work.